typing/types.ml

```diff
@@ -996,19 +1019,135 @@ end
 let eq_type t1 t2 = t1 == t2 || repr t1 == repr t2
 let compare_type t1 t2 = compare (get_id t1) (get_id t2)
 
+(* with-bounds *)
+
+(* Compare types roughly semantically, to allow best-effort deduplication of the types
+   inside of with-bounds.
+
+   This function might compare two types as inequal that are actually equal, but should
+   /never/ compare two types as equal that are not semantically equal. It may go without
+   saying but it also needs to expose a total order.
+
+   Someday, it's probably desirable to merge this, and make it compatible, with
+   [Ctype.eq_type], though that seems quite hard.
+*)
+let rec best_effort_compare_type_expr te1 te2 =
+  if te1 == te2 || repr te1 == repr te2 then 0
+  else
+    let rank ty =
+      match ty.desc with
+      (* Types which must be compared by id *)
+      | Tvar _
+      | Tobject (_, _)
+      | Tfield (_, _, _, _)
+      | Tnil
+      | Tlink _
+      | Tsubst (_, _)
+      | Tvariant _
+      | Tpackage (_, _)
+      | Tarrow (_, _, _, _)
```

Comment:
`Tarrow` seems easy enough (and prevalent enough) to do a more proper treatment for -- unless there's a detail making it harder than it looks.

Reply:
`commutable` seemed somewhat intimidating to get correct. Do you think it's easier than I'm expecting?